### ISSUE TITLE
Add cross-compiled library to RIOT OS

### ISSUE THREAD

**Issue by forellendev** [RIOT-OS/RIOT#16799]
Hi guys,

I tried adding a larger library to RIOT OS, but found that it isn't optimized enough. I compiled the library for the appropriate device architecture that I'm using RIOT on, but I have no idea how to add the compiled library to the project? Should I use PKG or MODULE in this case?

Best regards

**Comment by miri64** [comment]
Hi @forellendev, have a look at https://github.com/RIOT-OS/RIOT/blob/master/examples/bindist/README.md. I think that fits your use case.

**Comment by forellendev** [comment]
Thank you @miri64, I'll try it out!

**Comment by forellendev** [comment]
Unfortunately, I get the following error:
```
The selected TOOLCHAIN=/usr/share/gcc-arm-none-eabi-10.3-2021.07/bin is not supported.
Supported toolchains: gnu llvm afl
```
I cross-compiled a library on my ubuntu-x86_64 system to the ARM Cortex-M4 with the **arm-none-eabi toolchain** and would like to use this cross-compiled library within RIOT OS. Is this actually possible with the **arm-non-eabi** toolchain?

**Comment by miri64** [comment]
You don't need to set the toolchain. It just need to be in your `PATH` (which for a selected `BOARD` is usually the case). Just use a Cortex-M4-based board for testing.

**Comment by forellendev** [comment]
Thank you @miri64, I forgot to outcomment a path variable. I thought a little about manual cross-compiling the library, but now want to try to add it as a package (and hopefully as a new pkg to RIOT OS). I have the following Makefile in a new pkg folder:

```
PKG_NAME = libsafecrypto
PKG_URL = https://github.com/safecrypto/libsafecrypto.git
PKG_VERSION = 65b962f6d608704336ac16f226c8a5ff63b75851 # 0.2.0-52
PKG_LICENSE = MIT

include $(RIOTBASE)/pkg/pkg.mk

.PHONY: libsafecrypto%

LIBSAFECRYPTO_MODULES = \
INCLUDES += -I$(PKGDIRBASE)/libsafecrypto/include \
INCLUDES += -I$(PKGDIRBASE)/libsafecrypto/bindings \
INCLUDES += -I$(PKGDIRBASE)/libsafecrypto/build_tools \
INCLUDES += -I$(PKGDIRBASE)/libsafecrypto/src \
INCLUDES += -I$(PKGDIRBASE)/libsafecrypto/src/utils/crypto \
INCLUDES += -I$(PKGDIRBASE)/libsafecrypto/src/utils/threading \
INCLUDES += -I$(PKGDIRBASE)/libsafecrypto/src/schemes \

all:
	$(MAKE) -C $(PKG_SOURCE_DIR) -f $(RIOTBASE)/Makefile.base
```

The problem is that nothing is downloaded at all and my bin/native/libsafecrypto folders are empty.

**Comment by miri64** [comment]
Did you include `USEPKG += libsafecrypto` into your test application's Makefile?

**Comment by forellendev** [comment]
I did add this command to the Make of the application. Do you think it would be possible to fetch the project via PKG_URL and PKG_VERSION, but then apply a custom cross-compilation script to this download next? Maybe this library leads to some troubles.

**Comment by miri64** [comment]
> but then apply a custom cross-compilation script to this download next? Maybe this library leads to some troubles.

Yes, using custom compile scripts is what we often do e.g. with `pkg`s that use `cmake` or other non-`make` tools for building. Have a look e.g. at [`ccn-lite`](https://github.com/RIOT-OS/RIOT/blob/bfe64fe66a163e506824ea8d4ce30691dff78535/pkg/ccn-lite/Makefile), how the building step is handled there:

https://github.com/RIOT-OS/RIOT/blob/bfe64fe66a163e506824ea8d4ce30691dff78535/pkg/ccn-lite/Makefile#L22-L31

Basically, the make targets are the files you want to have generated, the command section of those targets provide said script (you can also call external scripts there) provide the instructions how this file will be generated.

**Comment by miri64** [comment]
@forellendev did my last comment solve your problem?

**Comment by maribu** [comment]
Let's assume the problem has been solved. If not, please reopen the issue - we are happy to help.
